README.md
# QualWeb core

The core allows you to perform automatic accessibility evaluations on web pages. It contains 3 evaluation modules:

- [@qualweb/act-rules](https://github.com/qualweb/act-rules)
- [@qualweb/wcag-techniques](https://github.com/qualweb/wcag-techniques)
- [@qualweb/best-practices](https://github.com/qualweb/best-practices)

You can also perform evaluations at [http://qualweb.di.fc.ul.pt/evaluator/](http://qualweb.di.fc.ul.pt/evaluator/), or by installing the [chrome extension](https://chrome.google.com/webstore/detail/qualweb-extension/ljgilomdnehokancdcbkmbndkkiggioc).

## How to install

```shell
  $ npm i @qualweb/core --save
```

## How to run

```javascript
  'use strict';

  const { QualWeb, generateEARLReport } = require('@qualweb/core');

  (async () => {
    const plugins = {
      // Check https://github.com/berstend/puppeteer-extra/tree/master/packages/puppeteer-extra-plugin-adblocker
      adBlock: true, // Default value = false
      // Check https://github.com/berstend/puppeteer-extra/tree/master/packages/puppeteer-extra-plugin-stealth
      stealth: true // Default value = false
    };
    const qualweb = new QualWeb(plugins);

    const clusterOptions = {
      maxConcurrency: 5, // Performs several urls evaluations at the same time - the higher the number given, more resources will be used. Default value = 1
      timeout: 60 * 1000, // Timeout for loading page. Default value = 30 seconds
      monitor: true // Displays urls information on the terminal. Default value = false
    };

    const launchOptions = {
      ... // check https://github.com/puppeteer/puppeteer/blob/v10.1.0/docs/api.md#puppeteerlaunchoptions
      // In most cases there's no need to give additional options. Just leave the field undefined
    };

    // Starts the QualWeb core engine
    await qualweb.start(clusterOptions, launchOptions);

    // QualWeb evaluation report
    const qualwebOptions = {
      url: 'https://act-rules.github.io/pages/about/',
      ...
    };

    // Evaluates the given options - will only return after all urls have finished evaluating or resulted in an error
    const reports = await qualweb.evaluate(qualwebOptions);

    console.log(reports);
    //  {
    //    "url": "report",
    //    "url2": "report2"
    //  }

    // Stops the QualWeb core engine
    await qualweb.stop();

    const earlOptions = {
      // Check the options in the section below
    };

    // if you want an EARL report
    const earlReports = generateEARLReport(reports, earlOptions);

    console.log(earlReports);
    //  {
    //    "url": "earlReport",
    //    "url2": "earlReport2"
    //  }
  })();
```

## Options

The available options fot the **evaluate()** function are:

```jsonc
{
  "url": "https://act-rules.github.io/pages/about/", // url to evaluate
  "urls": ["https://act-rules.github.io/pages/about/", "https://act-rules.github.io/rules/"], // Array of urls
  "file": "/path/to/file/with/urls", // urls must be separated by a newline (\n)
  "crawl": "https://act-rules.github.io", // Domain to crawl and obtain the urls
  "qualstate": {
    "waitTime": 2000, // time to wait between actions
    "maxStates": 2, // number of maximum states to be found
    "numberOfProcess": 3, // Number of concurrent processes. Default value = 1
    "log": {
      "file": true, // Logs errors to a file. Default value = false
      "console": false // Logs errors to the console. Default value = false
    },
    "viewport": {
      "mobile": false, // default value = false
      "landscape": true, // default value = viewPort.width > viewPort.height
      "resolution": {
        "width": 1920, // default value for desktop = 1366, default value for mobile = 1080
        "height": 1080 // default value for desktop = 768, default value for mobile = 1920
      }
    },
    "ignore": {
      "ids_compare": ["pErrorMessage"], // array of IDs or any identifying attribute of element that should be ignore when comparing states.
      "ids_events": ["btnReq", "hoverTest", "btnId", "aHref", "idTd1", "idTd2"] // array of IDs of element that have events and should be ignore.
    },
    "interaction": {
      "inputs": [ // array of inputs, that will be executed on the page if the inputs exist
        {
          "value": { // one or more key:value, where key is the ID of input and value the input value
            "onChangeInput": "inputOnChange"
          },
          "info": {
            "wait": 2000 // time to wait after executing the inputs
          }
        }
      ],
      "forms": [ // array of forms, that will be executed on the page if the forms exist
        {
          "input": [ // array inputs to be place on the form
            { // one or more key:value, where key is the ID of input and value the input value
              "fname": "filipe"
            }
          ],
          "action": { // action that submits the form
            "id": "btnSubmit", // ID of the element
            "event": "click" // type of event to be executed
          },
          "info": {
            "formId": "idForm" // ID of the form
          },
        }
      ],
      "directions": [ // array of directions, that will be executed at the beginning of the crawl process
        {
          "actions": [ // array of actions to be executed for each directions
            {
              "values": { // one or more key:value, where key is the ID of input and value the input value
                "fname": "filipe"
              },
              "action": { // action that should be trigger
                "id": "btnSubmit", // ID of the element
                "eventType": "click" // type of event to be executed
              }
            },
            {
              "values": {
                "newFName": "filipe"
              },
              "action": {
                "id": "btnSubmitNewForm",
                "eventType": "click"
              }
            }
          ],
          "info": {
            "crawl": "stop", // flag to know if the crawl process shoulf continue after the directions process. Value = "stop" || "continue"
            "save": true // save the states during the directions process
          }
        }
      ]
    }
  },
  "html": "<html-code>", // Full webpage html, or just small snippets
  "log": {
    "file": true, // Logs errors to a file. Default value = false
    "console": false // Logs errors to the console. Default value = false
  },
  "viewport": {
    "mobile": false, // default value = false
    "landscape": true, // default value = viewPort.width > viewPort.height
    "userAgent": "custom user agent", // default value for desktop = 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10.8; rv:22.0) Gecko/20100101 Firefox/22.0', default value for mobile = 'Mozilla/5.0 (Linux; U; Android 2.2; en-us; DROID2 GLOBAL Build/S273) AppleWebKit/533.1 (KHTML, like Gecko) Version/4.0 Mobile Safari/533.1'
    "resolution": {
      "width": 1920, // default value for desktop = 1366, default value for mobile = 1080
      "height": 1080 // default value for desktop = 768, default value for mobile = 1920
    }
  },
  "waitUntil": ["load", "networkidle0"], // Events to wait before starting evaluation, default value = "load". For more check https://github.com/puppeteer/puppeteer/blob/v10.1.0/docs/api.md#pagegotourl-options
  "validator": "http://127.0.0.1/validate", // HTML validator service endpoint. The url will be attached after of the given endpoint
  "crawlOptions": {
    "maxDepth": 2, // max depth to search, 0 to search only the given domain. Default value = -1 (search everything)
    "maxUrls": 100, // max urls to find. Default value = -1 (search everything)
    "timeout": 60, // how many seconds the domain should be crawled before it ends. Default value = -1 (never stops)
    "maxParallelCrawls": 10, // max urls to crawl at the same time. Default value = 5
    "logging": true // logs domain, current depth, urls found and time passed to the terminal
  },
  "translate": "en", // OR { "translate": "en", "fallback": "en" } OR { own translation object } check https://github.com/qualweb/locale#readme. Default = "en"
  "execute": {
    // choose which modules to execute
    // if this option is not specified, the default values below will be applied, otherwise unspecified values will default to false
    "wappalyzer": false, // wappalyzer module (https://github.com/qualweb/wappalyzer)
    "act": true, // act-rules module (https://github.com/qualweb/act-rules)
    "wcag": true, // wcag-techniques module (https://github.com/qualweb/wcag-techniques)
    "bp": false, // best-practices module (https://github.com/qualweb/best-practices)
    "counter": false // counter module (https://github.com/qualweb/counter)
  },
  "act-rules": {
    // More information about this options at https://github.com/qualweb/act-rules
    "rules": ["QW-ACT-R1", "b5c3f8"], // Array of rules to execute, can be passed the QualWeb Rule ID or the ACT Rule ID
    "exclude": ["QW-ACT-R1", "b5c3f8"], // Array of rules to exclude, can be passed the QualWeb Rule ID or the ACT Rule ID
    "levels": ["A", "AA", "AAA"], // Conformance levels to execute,
    "principles": ["Perceivable", "Operable", "Understandable", "Robust"] // Principles to execute
  },
  "wcag-techniques": {
    // More information about this options at https://github.com/qualweb/wcag-techniques
    "techniques": ["QW-WCAG-T1", "H39"], // Array of techniques to execute, can be passed the QualWeb Technique ID or the WCAG 2.1 Technique Code
    "exclude": ["QW-WCAG-T1", "H39"], // Array of techniques to exclude, can be passed the QualWeb Technique ID or the WCAG 2.1 Technique Code
    "levels": ["A", "AA", "AAA"], // Conformance levels to execute,
    "principles": ["Perceivable", "Operable", "Understandable", "Robust"] // Principles to execute
  },
  "best-practices": {
    // More information about this options at https://github.com/qualweb/best-practices
    "bestPractices": ["QW-BP1", "QW-BP2"], // Array of best practices to execute
    "exclude": ["QW-BP1", "QW-BP2"] // Array of best practices to exclude
  }
}
```

The available options fot the **generateEARLReport()** function are:

```jsonc
{
  "aggregated": true, // default value = false
  "aggregatedName": "websites.json", // The name to save the aggregated earl reports. default value = first url of the list
  "modules": {
    // Choose which modules to convert the report to earl, by default all modules are converted if they were executed
    "act": true, // default value = true
    "wcag": false, // default value = true
    "best-practices": false // default value = true
  } // If the "modules" value is given, any missing module value missing it's automatically set to false
}
```

## Report details

In this section it's explained the evaluation report in detail. For a detailed version of the EARL report check [@qualweb/earl-reporter](https://github.com/qualweb/earl-reporter#report-details).

```jsonc
  {
    "type": "evaluation",
    "system": {
      "name": "QualWeb",
      "description": "QualWeb is an automatic accessibility evaluator for webpages.",
      "version": "QualWeb version",
      "homepage": "http://www.qualweb.di.fc.ul.pt/",
      "date": "date of the evaluation",
      "hash": "unique hash",
      "url": {
        "inputUrl": "inserted url",
        "protocol": "protocol of the url",
        "domainName": "domain name of the url",
        "domain": "domain of the url",
        "uri": "uri of the url",
        "completeUrl": "complete url after all redirects"
      },
      "page": {
        "viewport": {
          "mobile": "was evaluated on a mobile context or not",
          "landscape": "was evaluated on a landscape context or not",
          "userAgent": "user agent used",
          "resolution": {
            "width": "window's width used",
            "height": "window's height used",
          }
        },
        "dom": {
          "html": "html code as a string",
          "title": "Title of the webpage",
          "elementCount": "Element count of the webpage"
        }
      }
    },
    "metadata": {
      "passed": "number of passed rules/techniques/best practices",
      "warning": "number of warning rules/techniques/best practices",
      "failed": "number of failed rules/techniques/best practices",
      "inapplicable": "number of inapplicable rules/techniques/best practices",
    },
    "modules": {
      "act-rules": {
        "type": "act-rules",
        "metadata": {
          "passed": "number of passed rules",
          "warning": "number of warning rules",
          "failed": "number of failed rules",
          "inapplicable": "number of inapplicable rules",
        },
        "assertions": {
          "QW_ACT_R1": {
            "name": "Name of the rule",
            "code": "QualWeb rule id",
            "mapping": "ACT rule id mapping",
            "description": "Description of the rule",
            "metadata": {
              "target": "Any target, can be one element, multiple elements, attributes, a relation between elements",
              "success-criteria?": [
                {
                  "name": "Name of the success criteria",
                  "level": "Level of conformance of the success criteria",
                  "principle": "Principle of the success criteria",
                  "url": "Url of the success criteria"
                }
              ],
              "related?": [], // related WCAG 2.1 techniques
              "url?": "Url of the rule",
              "passed": "Number of passed results",
              "warning": "Number of warning results",
              "failed": "Number ff failed results",
              "type?": [], // usually "ACTRule" or "TestCase"
              "a11yReq?": [], // WCAG 2.1 relation - something like "WCAG21:language"
              "outcome": "Outcome of the rule",
              "description": "Description of the outcome";
            },
            "results": [
              {
                "verdict": "Outcome of the test",
                "description": "Description of the test",
                "resultCode": "Test identifier",
                "pointer?": "Element pointer in CSS notation",
                "htmlCode?": "Element html code",
                "attributes?": "Attributes of the element",
                "accessibleName?": "Accessible name of the test target"
              },
              { ... }
            ]
          },
          "...": { ... }
        }
      },
      "wcag-techniques": {
        "type": "wcag-techniques",
        "metadata": {
          "passed": "number of passed techniques",
          "warning": "number of warning techniques",
          "failed": "number of failed techniques",
          "inapplicable": "number of inapplicable techniques",
        },
        "assertions": {
          "QW_WCAG_T1": {
            "name": "Name of the technique",
            "code": "QualWeb technique id",
            "mapping": "WCAG techniques code mapping",
            "description": "Description of the technique",
            "metadata": {
              "target": "Any target, can be one element, multiple elements, attributes, a relation between elements",
              "success-criteria?": [
                {
                  "name": "Name of the success criteria",
                  "level": "Level of conformance of the success criteria",
                  "principle": "Principle of the success criteria",
                  "url": "Url of the success criteria"
                }
              ],
              "related?": [], // related WCAG 2.1 techniques
              "url?": "Url of the technique",
              "passed": "Number of passed results",
              "warning": "Number of warning results",
              "failed": "Number ff failed results",
              "type?": [], // usually "ACTRule" or "TestCase"
              "a11yReq?": [], // WCAG 2.1 relation - something like "WCAG21:language"
              "outcome": "Outcome of the technique",
              "description": "Description of the outcome";
            },
            "results": [
              {
                "verdict": "Outcome of the test",
                "description": "Description of the test",
                "resultCode": "Test identifier",
                "pointer?": "Element pointer in CSS notation",
                "htmlCode?": "Element html code",

CHANGELOG.md
# Changelog

## [0.7.45] - 28/09/2023

### Updated

- add qualstate

## [0.7.44] - 28/09/2023

### Updated

- dependencies

## [0.7.43] - 27/09/2023

### Updated

- dependencies

## [0.7.42] - 26/09/2023

### Updated

- dependencies

## [0.7.41] - 19/09/2023

### Updated

- dependencies

## [0.7.40] - 07/09/2023

### Updated

- dependencies

## [0.7.39] - 05/09/2023

### Updated

- dependencies

## [0.7.38] - 04/09/2023

### Updated

- dependencies

## [0.7.37] - 18/07/2023

### Updated

- dependencies

## [0.7.35] - 07/06/2023

### Updated

- dependencies

## [0.7.34] - 30/05/2023

### Updated

- dependencies

## [0.7.33] - 08/03/2023

### Updated

- dependencies

## [0.7.32] - 01/02/2023

### Updated

- dependencies

## [0.7.31] - 17/01/2023

### Updated

- dependencies

## [0.7.30] - 17/01/2023

### Updated

- dependencies

## [0.7.29] - 11/01/2023

### Updated

- dependencies

## [0.7.28] - 29/06/2022

### Updated

- dependencies

## [0.7.27] - 28/06/2022

### Updated

- dependencies

## [0.7.26] - 26/05/2022

### Updated

- dependencies

## [0.7.25] - 26/05/2022

### Updated

- dependencies

## [0.7.24] - 23/05/2022

### Updated

- dependencies

## [0.7.23] - 23/05/2022

### Updated

- dependencies

## [0.7.22] - 19/05/2022

### Updated

- dependencies

## [0.7.21] - 12/04/2022

### Updated

- dependencies

## [0.7.20] - 28/03/2022

### Updated

- dependencies

## [0.7.19] - 21/02/2022

### Updated

- dependencies

## [0.7.18] - 01/02/2022

### Updated

- puppeteer from 13.1.2 to 13.1.3

## [0.7.17] - 26/01/2022

### Updated

- puppeteer from 10.4.0 to 13.1.2
- @qualweb/evaluation from 0.3.10 to 0.3.11

## [0.7.16] - 06/01/2022

### Updated

- dependencies

## [0.7.15] - 06/01/2022

### Updated

- dependencies

## [0.7.14] - 24/12/2021

### Updated

- dependencies

## [0.7.13] - 10/12/2021

### Updated

- dependencies

## [0.7.12] - 06/12/2021

### Updated

- @qualweb/locale

## [0.7.11] - 03/12/2021

### Updated

- dependencies

## [0.7.10] - 15/11/2021

### Updated

- dependencies

## [0.7.9] - 28/10/2021

### Updated

- @qualweb/crawler

## [0.7.8] - 21/10/2021

### Updated

- dependencies

## [0.7.7] - 21/10/2021

### Updated

- crawler method with viewport and waitUntil options
- dependencies

## [0.7.6] - 12/10/2021

### Updated

- dependencies

## [0.7.5] - 01/10/2021

### Updated

- dependencies

## [0.7.4] - 29/09/2021

### Added

- Finnish locale to @qualweb/locale package

### Updated

- dependencies

## [0.7.3] - 09/08/2021

### Added

- logging options (log files are now not generated by default)

## [0.7.2] - 28/07/2021

## BREAKING CHANGES

- result codes from evaluation modules have changed from "RCx" to "Px" (Passed), "Wx" (Warning), "Fx" (Failed) and "Ix" (Inapplicable)

### Updated

- dependencies

## [0.7.1-alpha] - 27/07/2021

### Updated

- localization support
- dependencies

## [0.6.14] - 23/07/2021

### Updated

- dependencies
- test files

## [0.6.14] - 22/07/2021

### Updated

- dependencies

## [0.6.13] - 21/07/2021

### Updated

- dependencies

## [0.6.12] - 30/06/2021

### Updated

- dependencies

## [0.6.11] - 21/06/2021

### Updated

- dependencies

## [0.6.10] - 14/06/2021

### Fixed

- error when using `html` option

### Updated

- dependencies

## [0.6.9] - 25/05/2021

### Updated

- dependencies

## [0.6.8] - 19/05/2021

### Updated

- dependencies

## [0.6.7] - 12/05/2021

### Updated

- dependencies

## [0.6.6] - 11/05/2021

### Changes

- qualweb error log file is now deleted if every evaluation succeeded

### Updated

- dependencies

## [0.6.5] - 10/05/2021

### Updated

- dependencies

## [0.7.0-alpha] - 05/05/2021

### Added

- localization support

## [0.6.4] - 03/05/2021

### Updated

- dependencies

## [0.6.3-alpha] - 27/04/2021

### Updated

- dependencies

## [0.5.7] - 27/04/2021

### Updated

- dependencies

## [0.5.6] - 23/04/2021

### Updated

- dependencies

## [0.6.2-alpha] - 26/04/2021

### Changed

- qualweb-errors.log to qualweb-errors-<timestamp>.log
  - now a new error.log file is created for every call of the evaluate() function

## [0.6.1-alpha] - 22/04/2021

### Updated

- @qualweb/types
- README.md

## [0.6.0-alpha] - 21/04/2021

### Added

- puppeteer-extra with AdBlocker and Stealth plugins (optional)
- puppeteer-cluster to enable faster evaluations of multiple webpages

### Updated

- puppeteer to version 9.0.0

## [0.5.5] - 15/04/2021

### Updated

- dependencies
- code optimizations

## [0.5.4] - 13/04/2021

### Updated

- dependencies
- documentation

## [0.5.3] - 01/04/2021

### Updated

- dependencies

## [0.5.2] - 31/03/2021

### Changes

- Removed QW-ACT-R8 since it was deprecated (https://act-rules.github.io/rules/9eb3f6)
- Changed dom data structure in the final evaluation report
  - now it only shows the processed dom information, instead of both source page and processed page
- Removed a lot of third party dependencies on all modules

### Updated

- dependencies

## [0.5.1] - 23/03/2021

### Added

- code documentation

### Updated

- earl-reporter
- @qualweb/types

## [0.5.0] - 23/03/2021

### Updated

- puppeteer to version 8.0.0

## [0.4.69] - 22/03/2021

### Updated

- dependencies
- exported functions

## [0.4.68] - 22/03/2021

### Updated

- dependencies

## [0.4.67] - 22/03/2021

### Updated

- dependencies

### Changed

## [0.4.66] - 12/03/2021

### Changed

- downgraded puppeteer from version 8.0.0 to 5.5.0 due to inconsistent certificate validation

## [0.4.65] - 11/03/2021

### Updated

- dependencies
- README.md

## [0.4.64] - 11/03/2021

### Updated

- dependencies
- README.md

## [0.4.63] - 10/03/2021

### Updated

- dependencies

## [0.4.62] - 08/03/2021

### Updated

- dependencies

## [0.4.61] - 08/03/2021

### Updated

- dependencies

## [0.4.60] - 03/03/2021

### Updated

- dependencies

## [0.4.59] - 03/03/2021

### Added

- element and role counter

## [0.4.58] - 01/03/2021

### Updated

- dependencies

## [0.4.57] - 27/02/2021

### Updated

- dependencies

## [0.4.56] - 26/02/2021


src/index.ts
import { LaunchOptions, BrowserLaunchArgumentOptions, BrowserConnectOptions, Viewport } from 'puppeteer';
import puppeteer from 'puppeteer-extra';
import { Cluster } from 'puppeteer-cluster';
import StealthPlugin from 'puppeteer-extra-plugin-stealth';
import AdBlocker from 'puppeteer-extra-plugin-adblocker';
import { QualwebOptions, Evaluations, PuppeteerPlugins, ClusterOptions, LoadEvent, QualwebPlugin, State } from '@qualweb/core';
import { generateEARLReport } from '@qualweb/earl-reporter';
import { Dom } from '@qualweb/dom';
import { Evaluation } from '@qualweb/evaluation';
import { Crawler, CrawlOptions } from '@qualweb/crawler';
import locales, { Lang, Locale, TranslationObject } from '@qualweb/locale';
import { readFile, writeFile, unlink } from 'fs';
import path from 'path';
import 'colors';


const qualstate = require("qualstate");

/**
 * QualWeb engine - Performs web accessibility evaluations using several modules:
 * - act-rules module (https://github.com/qualweb/act-rules)
 * - wcag-techniques module (https://github.com/qualweb/wcag-techniques)
 * - best-practices module (https://github.com/qualweb/best-practices)
 */
class QualWeb {
  /**
   * Chromium browser cluster.
   */
  private cluster?: Cluster;

  /**
   * Array of plugins added with QualWeb.use().
   */
  private qualwebPlugins: QualwebPlugin[] = [];

  /**
   * Initializes puppeteer with given plugins.
   *
   * @param {PuppeteerPlugins} plugins - Plugins for puppeteer - supported: AdBlocker and Stealth.
   */
  constructor(plugins?: PuppeteerPlugins) {
    if (plugins?.stealth) {
      puppeteer.use(StealthPlugin());
    }
    if (plugins?.adBlock) {
      puppeteer.use(AdBlocker({ blockTrackers: true }));
    }
  }

  /**
   * Starts chromium browser cluster.
   *
   * @param {ClusterOptions} clusterOptions - Options for cluster initialization.
   * @param {LaunchOptions & BrowserLaunchArgumentOptions & BrowserConnectOptions} puppeteerOptions - check https://github.com/puppeteer/puppeteer/blob/v9.1.1/docs/api.md#puppeteerlaunchoptions.
   */
  public async start(
    clusterOptions?: ClusterOptions,
    puppeteerOptions?: LaunchOptions & BrowserLaunchArgumentOptions & BrowserConnectOptions
  ): Promise<void> {
    this.cluster = await Cluster.launch({
      concurrency: Cluster.CONCURRENCY_CONTEXT,
      maxConcurrency: clusterOptions?.maxConcurrency ?? 1,
      puppeteerOptions: puppeteerOptions,
      puppeteer: puppeteer,
      timeout: clusterOptions?.timeout ?? 60 * 1000,
      monitor: clusterOptions?.monitor ?? false
    });
  }

  /**
   * Adds a plugin to be run when executing a QualWeb evaluation. Plugins are
   * called in the same order they were added to the instance.
   * @param plugin The plugin to add.
   * @returns The QualWeb instance itself. Good for chaining.
   */
  public use(plugin: QualwebPlugin): this {
    this.qualwebPlugins.push(plugin);

    return this;
  }

  /**
   * Closes chromium browser.
   */
  public async stop(): Promise<void> {
    await this.cluster?.close();
  }

  /**
   * Evaluates given options.
   *
   * @param {QualwebOptions} options - Options of execution (check https://github.com/qualweb/core#options).
   * @returns List of evaluations.
   */
  public async evaluate(options: QualwebOptions): Promise<Evaluations> {
    const modulesToExecute = {
      act: true,
      wcag: true,
      bp: true,
      wappalyzer: false,
      counter: false
    };

    const urls = await this.checkUrls(options);

    if ((options.html === undefined || options.html.trim() === '') && urls.length === 0) {
      throw new Error('Invalid input method');
    }

    this.handleLocales(options);

    if (options.execute) {
      modulesToExecute.act = !!options.execute.act;
      modulesToExecute.wcag = !!options.execute.wcag;
      modulesToExecute.bp = !!options.execute.bp;
      modulesToExecute.wappalyzer = !!options.execute.wappalyzer;
      modulesToExecute.counter = !!options.execute.counter;
    }

    let evaluations: Evaluations = {};

    let foundError = false;

    const timestamp = new Date().getTime();

    handleError(
      options,
      'Evaluation errors',
      new Date(timestamp).toISOString().replace(/T/, ' ').replace(/\..+/, '') + '\n-----------',
      timestamp
    );

    this.cluster?.on('taskerror', (err, data) => {
      foundError = true;
      handleError(options, data.url, err.message + '\n-----------', timestamp);
    });

    await this.cluster?.task(async ({ page, data: { url, html, stateData } }) => {
      const dom = new Dom(page, options.validator);

      // Execute beforePageLoad on all plugins in order. If any exceptions
      // occur during the execution of a plugin, it should bubble up past the
      // call to Qualweb.evaluate() so the user can handle it on their own.
      for (const plugin of this.qualwebPlugins) {
        if (typeof plugin.beforePageLoad === 'function') {
          await plugin.beforePageLoad(page, url || 'customHtml');
        }
      }

      const { sourceHtml, validation } = await dom.process(options, url ?? '', html ?? '');

      // Execute afterPageLoad on all plugins in order. Same assumptions for
      // exceptions apply as they did for beforePageLoad.
      for (const plugin of this.qualwebPlugins) {
        if (typeof plugin.afterPageLoad === 'function') {
          await plugin.afterPageLoad(page, url || 'customHtml');
        }
      }

      const evaluation = new Evaluation(url, page, modulesToExecute);
      const evaluationReport = stateData != null ?
        await evaluation.evaluatePage(sourceHtml, options, validation, stateData.selector) :
        await evaluation.evaluatePage(sourceHtml, options, validation);

      if (stateData != null) {
        stateData.arrStates.push(evaluationReport.getFinalReport());
      } else {
        evaluations[url || 'customHtml'] = evaluationReport.getFinalReport();
      }
    });

    let statesArr: State[] = [];

    if (options.qualstate != null) {
      try {
        options.qualstate["url"] = urls[0];
        let states = await qualstate.crawl(options.qualstate, puppeteer); // try
        for (const state of states) {
          let stateData = {
            selector: state._selector,
            arrStates: statesArr
          }
          console.log("States Size: "+ states.size);
          this.cluster?.queue({ html: state._body, stateData: stateData });
        }
      } catch (error) {
        throw new Error('Error: ' + error);
      }
    } else {
      for (const url of urls) {
        this.cluster?.queue({ url });
      }

      if (options.html) {
        this.cluster?.queue({ html: options.html });
      }
    }

    await this.cluster?.idle();

    if (options.qualstate != null) {
      const browser = await puppeteer.launch();
      const page = await browser.newPage();
      const evaluationQS = new Evaluation(urls[0], page, modulesToExecute);
      const evaluationReport = await evaluationQS.evaluatePage('', options, undefined, undefined, true);
      evaluations[urls[0]] = evaluationReport.getFinalReportQS();
      evaluations[urls[0]]["states"] = statesArr as [State];
    }

    if (options.log?.file) {
      if (foundError) {
        console.warn('One or more urls failed to evaluate. Check the error.log for more information.'.yellow);
      } else {
        deleteErrorLogFile(timestamp);
      }
    }

    return evaluations;
  }

  /**
   * Crawls a webpage to find all urls.
   *
   * @param {string} startingUrl - Webpage to crawl.
   * @param {CrawlOptions} options - Options for crawling process.
   * @Param {Viewport} viewport - Set the viewport of the webpages.
   * @param {LoadEvent | Array<LoadEvent>} waitUntil - Wait for dom events before starting the crawling process.
   * @returns List of decoded urls.
   */
  public async crawl(
    domain: string,
    options?: CrawlOptions,
    viewport?: Viewport,
    waitUntil?: LoadEvent | Array<LoadEvent>
  ): Promise<Array<string>> {
    const browser = await puppeteer.launch();
    const incognito = await browser.createIncognitoBrowserContext();
    const crawler = new Crawler(incognito, domain, viewport, waitUntil);
    await crawler.crawl(options);
    return crawler.getResults();
  }

  /**
   * Checks possible input options and compiles the urls.
   * Possible input options are:
   * - url - single url;
   * - urls - multiple urls;
   * - filepath - path to file with urls;
   * - crawler - domain to crawl and gather urls.
   *
   * @param {QualwebOptions} options - QualWeb options.
   * @returns List of urls.
   */
  private async checkUrls(options: QualwebOptions): Promise<Array<string>> {
    const urls = new Array<string>();
    if (options.url) {
      urls.push(decodeURIComponent(options.url).trim());
    }
    if (options.urls) {
      urls.push(...options.urls.map((url: string) => decodeURIComponent(url).trim()));
    }
    if (options.file) {
      urls.push(...(await getFileUrls(options.file)));
    }
    if (options.crawl) {
      const viewport = {
        width: 0,
        height: 0,
        isMobile: false,
        isLandscape: true
      };
      if (options.viewport) {
        viewport.width = options?.viewport?.resolution?.width ?? 0;
        viewport.height = options?.viewport?.resolution?.height ?? 0;
        viewport.isMobile = options?.viewport?.mobile ?? false;
        viewport.isLandscape = options?.viewport?.landscape ?? true;
      }
      urls.push(
        ...(await this.crawl(
          options.crawl,
          options.crawlOptions,
          viewport.width + viewport.height !== 0 ? viewport : undefined,
          options.waitUntil
        ))
      );
    }

    return urls;
  }

  private handleLocales(options: QualwebOptions): void {
    if (options.translate) {
      if (typeof options.translate === 'string') {
        if (Object.keys(locales).includes(options.translate)) {
          options.translate = {
            translate: locales[<Lang>options.translate],
            fallback: locales.en
          };
        } else {
          throw new Error(`Locale "${options.translate}" not supported.`);
        }
      } else if (Object.keys(options.translate).includes('translate')) {
        this.verifyTranslationObject(options);
      } else {
        options.translate = {
          translate: <Locale>options.translate,
          fallback: locales.en
        };
      }
    } else {
      options.translate = {
        translate: locales.en,
        fallback: locales.en
      };
    }
  }

  private verifyTranslationObject(options: QualwebOptions): void {
    if (typeof (<TranslationObject>options.translate).translate === 'string') {
      if (Object.keys(locales).includes(<string>(<TranslationObject>options.translate).translate)) {
        options.translate = {
          translate: locales[<Lang>(<TranslationObject>options.translate).translate],
          fallback: locales.en
        };
      } else {
        throw new Error(`Locale "${(<TranslationObject>options.translate).translate}" not supported.`);
      }
    }
    if (typeof (<TranslationObject>options.translate).fallback === 'string') {
      if (Object.keys(locales).includes(<string>(<TranslationObject>options.translate).fallback)) {
        (<TranslationObject>options.translate).fallback =
          locales[<Lang>(<TranslationObject>options.translate).fallback];
      } else {
        throw new Error(`Locale "${(<TranslationObject>options.translate).fallback}" not supported.`);
      }
    }
  }
}

/**
 * Reads a file to obtain the urls to evaluate.
 *
 * @param {string} file - Path to file of urls.
 * @returns List of decoded urls.
 */
async function getFileUrls(file: string): Promise<Array<string>> {
  const content = await readFileData(file);
  return content
    .split('\n')
    .map((url: string) => {
      try {
        return decodeURIComponent(url).trim();
      } catch (_err) {
        return '';
      }
    })
    .filter((url: string) => url.trim() !== '');
}

/**
 * Reads a file.
 *
 * @param {string} file - Path to file.
 * @returns File data converted to UTF-8.
 */
function readFileData(file: string): Promise<string> {
  return new Promise((resolve, reject) => {
    readFile(file, (err, data) => {
      if (err) {
        reject(err);
      } else {
        resolve(data.toString('utf-8'));
      }
    });
  });
}

/**
 * Logs evaluation errors to a error log file.
 *
 * @param {options} QualwebOptions - options to check logging method
 * @param {string} url - Url that failed to evaluate.
 * @param {string} message - Error message of the evaluation.
 * @param {number} timestamp - Date of the evaluation.
 */
function handleError(options: QualwebOptions, url: string, message: string, timestamp: number): void {
  if (options.log && options.log.file) {
    writeFile(
      path.resolve(process.cwd(), `qualweb-errors-${timestamp}.log`),
      url + ' : ' + message + '\n',
      { flag: 'a', encoding: 'utf-8' },
      (err) => {
        if (err) {
          console.error(err);
        }
      }
    );
  }
  if (options.log && options.log.console) {
    console.error(url + ' : ' + message + '\n');
  }
}

/**
 * Deletes the error log file created at the beginning of the evaluation if there aren't any errors to report.
 *
 * @param {number} timestamp - Date of the evaluation.
 */
function deleteErrorLogFile(timestamp: number): void {
  unlink(path.resolve(process.cwd(), `qualweb-errors-${timestamp}.log`), (err) => {
    if (err) {
      throw err;
    }
  });
}

export { QualWeb, generateEARLReport, getFileUrls };
